server/portal/apps/auth/models.py

```diff
@@ -79,16 +81,14 @@ def client(self):
                        refresh_token=self.refresh_token)
 
         with transaction.atomic():
-            if self.expired:
-                try:
-                    client.refresh_tokens()
-                except Exception:
-                    logger.exception('Tapis Token refresh failed')
-                    raise
-
-                self.update(created=int(time.time()),
-                            access_token=client.access_token.access_token,
-                            expires_in=client.access_token.expires_in().total_seconds())
+            try:
+                if self.expired:
+                    with self.lock:
+                        if self.expired:
```

Comment:
looks good but something ('refresh_from_db' ?) is needed to ensure that the model's `expires_in` and `created` have up-to-date values after the lock is acquired.

Reply:
I used refresh_from_db and requests waiting for the lock are still seeing old data. 

Apart from that, I tested this by adding log lines. We will need unit tests in this area.
1) Test with no expiry, nothing is updated in db. 
2) Test with forced expiry of token. Token is refreshed but db is not updated.

Second test failed. Below update is needed to fix that.
```
if self.expired:
                with self.lock:
                    if self.expired:
                        try:
                            client.refresh_tokens()
                        except Exception:
                            logger.info('Tapis Token refresh failed User id %d', self.user.id)
                            raise
                        self.update(created=int(time.time()), access_token=client.access_token.access_token, expires_in=client.access_token.expires_in().total_seconds())
                        self.refresh_from_db()
```
There are 3 adjustments needed to make the test pass:
* try - catch is specific to client.refresh_tokens() (keeps this in par of existing logic)
* self.update is needed to update db
* self.refresh_from_db to update model.

Even with above code, the multi thread scenario still requested for token. The failing test case:
* force expiry
* Refresh browser (sends multiple actions at once).
* Each action is refreshing token (see created time in db, the token creation time keeps updating). 


The refresh_from_db does not update waiting threads. I'm looking further.

Reply:
Note: I used the refresh_from_db at wrong place, it should be before self.expired. Trying that out.